Nullables with proto3 using Google wrappers

jkuklis opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Nullables detection in protobuf using Google wrappers

## Description

In proto3, default values are not sent on the wire. This makes it non-trivial to distinguish between null and default values for Nullable columns.

A standard way to deal with this problem is to use Google wrappers to nest the target value within an inner message (see https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/wrappers.proto). In this case, a missing field is interpreted as null value, a field with missing value if interpreted as default value, and a field with regular value is interpreted as regular value.

However, ClickHouse interprets Google wrappers as nested columns. We propose to introduce special behaviour to detect Google wrappers and interpret them like in the description above. For example, to serialize values for a Nullable column `test`, we would use `google.protobuf.StringValue test` in our .proto schema. Note that these types are so called "well-known types" in Protobuf, implemented in the library itself.

The current implementation changes only a single .cpp file. Detecting that a given column is Nullable to insert the nested defaults is tricky here (see lines 2306 and onwards). Another design would be to add method `insertNonNullDefaults` to `ProtobufSerializer`, implemented as a call to `insertDefaults`, and overriden in `ProtobufSerializerNullable` to do what `insertNestedDefaults` does now. This would remove the need to detect if the column is Nullable.

We have only implemented the read part so far, but a similar pattern could be applied to the write part as well.

If this change was to break backward compatibility, the special treatment of Google wrappers could be enabled with an extra setting.

Let us know what you think about this approach.
Thanks!

Jakub Kuklis
Contentsquare

jkuklis committed:
Nullables detection in protobuf using Google wrappers

jkuklis committed:
Style corrections

jkuklis committed:
Fixing assert

jkuklis committed:
const keyword corrections for debug build

jkuklis committed:
Another const keyword corrections for debug build

antonio2368 commented:
I like the idea but I'm not sure about all the implications of it.

Just to make it clear, the idea is to have values wrapped with Google wrappers as nullable versions of them. E.g. `DoubleValue` in proto is a `nullable` `double` column while `double` in proto is a nonnullable `double` column?

That would also mean that we would need to write all the nullable columns using those wrappers. 
Based on the documentation itself it seems that the wrappers really are used almost exclusively for the use-case you described, to distinguish between null and default.

Nevertheless, I would still implement this as an opt-in feature with a setting, and leave the default behaviour as is because I think it can be confusing if we use `DoubleValue` for one case and `double` for another implicitly.

antonio2368 commented:
I like the idea but I'm not sure about all the implications of it.

Just to make it clear, the idea is to have values wrapped with Google wrappers as nullable versions of them. E.g. `DoubleValue` in proto is a `nullable` `double` column while `double` in proto is a nonnullable `double` column?

That would also mean that we would need to write all the nullable columns using those wrappers. 
Based on the documentation itself it seems that the wrappers really are used almost exclusively for the use-case you described, to distinguish between null and default.

Nevertheless, I would still implement this as an opt-in feature with a setting, and leave the default behaviour as is because I think it can be confusing if we use `DoubleValue` for one case and `double` for another implicitly.

jkuklis commented:
@antonio2368,

To be clear, the current implementation in the PR works as follows:
* Google wrapper type used for a nullable column - interpret lack of wrapper as null, wrapper with missing value as default, and wrapper with value present as real value,
* Google wrapper type used for a regular column - interpret lack of wrapper / wrapper with missing value as default, and wrapper with value present as real value,
* Regular type used for a nullable column - no changes: interpret lack of value as null, and value present as real value,
* Regular type used for a regular column - no changes.

In a way, in the read pipeline there are no other changes to be made: using Google wrappers is already an opt-in in this case, the behaviour of regular types doesn't change, and they can still be used for nullable columns like before (without a natural way to differentiate between null and default values). A setting doesn't seem necessary for this part: I think the only way Google wrappers could be used before would be for nested columns with attribute name "value", but I don't see any reason for using Google wrappers in this case, and the current implementation handles such cases well anyway (or at least could easily).

The writing part is kind of independent on the read pipeline, so technically Google wrappers could have this special treatment only in the read pipeline (as regular types can still be used for nullable columns like before). This seems rather inconsistent, so indeed it would likely be better to at least have an option to write nullable columns as Google wrapper types. I agree that it would be confusing if Google wrapper types were enforced for nullable columns, so a setting in this part is necessary - if, of course, Google wrappers are used at all in this part.

antonio2368 commented:
>  I think the only way Google wrappers could be used before would be for nested columns with attribute name "value"

I was thinking just to eliminate this case, because it would be strange if we handle some messages in a special way.
Even though it doesn't make much sense to store as a Nested type, it's still aligned with how other messages are handled so it's less unexpected.

jkuklis committed:
Write empty values for Google wrappers

jkuklis committed:
Test for input and output of Nullable Google wrappers, still one thing to investigate about output

jkuklis committed:
Corrected the behaviour for Proto Nullable output

jkuklis committed:
Remove done TODO

jkuklis committed:
Alternative design for output, mory messy, but the default value inside Google wrapper is not serialized

jkuklis committed:
Adding a setting to enable special treatment of google wrappers

jkuklis committed:
Adding a setting to enable Google wrappers special treatment

jkuklis committed:
Remove local change

jkuklis committed:
Improve the test clarity

jkuklis committed:
Add a test with multiple wrappers used in a single message

jkuklis committed:
Minor improvements

jkuklis commented:
Hello @antonio2368,

I needed to switch to a different project for some time, but I am back on this topic.
I have introduced the changes that we discussed: a setting to enable special treatment of Google wrappers and the write part modifications. I added a correctness test as well, the code seems to work as intended.

I have some remarks and questions regarding the design and the test:
* The only way to use Google value wrappers in the input pipeline without my modifications is for a `Tuple` with nested column named `value`. With my modifications, this use case would not crash, but the value would not be serialized correctly, it would always remain default. I tried to make both approaches work at the same time, I was pretty close, but this required recognizing whether the column we serialize has nested fields at `setColumn`, and additional handling to maintain this information. It was rather messy, and this previous use case is quite obscure, so I decided to just introduce a setting for my modifications, as you asked for it anyway.
* Regarding the output pipeline and your concerns, the setting also removes all interference here, but in this case it is up to the user whether he wants to serialize his column as either a regular protobuf value or a Google value wrapper - it is specified in the proto schema that is passed with the query. In this respect, there shouldn't be any surprises to the ClickHouse users, their values won't be serialized into a different type than before if this change is introduced. The only difference in the current implementation is that Google wrappers carrying default values will be serialized as empty wrappers, as intended.
* Should the setting be `true` by default? I think it would be better this way, as I don't think Google value wrappers were used frequently before due to this lack of use cases, maybe only for some compatibility with other tools.
* I didn't come up with a name of the setting better than `format_protobuf_google_wrappers_special_treatment`, I am also unsure how to concisely describe its meaning. The idea is that we can use Google value wrappers (which are nested messages in protobuf) for regular non-nested ClickHouse columns.
* At the moment the `wrappers.proto` file is copy-pasted from https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/wrappers.proto. It is actually available in the ClickHouse `contrib/protobuf` directory, but protobuf imports do not allow characters like `.` in the filenames, so I couldn't use a path starting with `../`. From https://github.com/ClickHouse/ClickHouse/issues/11110 and what I've seen the import is on the client side, so the import must be a relative path. Do you know if it is possible to handle it without a copy-paste?
* The test I introduced passes locally, but in the pipeline in the PR I get the following error: `Code: 36. DB::Exception: The format Protobuf requires a schema. The corresponding setting should be set: data for INSERT was parsed from query: (in query: INSERT INTO roundtrip_google_wrappers_02266 FORMAT Protobuf SETTINGS format_schema='/usr/share/clickhouse-test/queries/0_stateless/format_schemas/02266_protobuf_format_google_wrappers:Message'). (BAD_ARGUMENTS)`. So the error description says that the schema is not present, but in the query it is present. This happens the second time this schema is used, the first time for input - the first time the schema is used passes for output. Do you know what might cause this error?
* In the current approach default value wrappers are serialized as empty wrappers. I first started with an alternative approach, in which default value wrappers are serialized as wrappers with the default value serialized as well, I roughly finished this approach in commit [7713dba](https://github.com/ClickHouse/ClickHouse/pull/35149/commits/7713dba1b9f66e2baa9c90397b3de330dbbbeefe). There was less code there, but it was before I introduced the settting, which for this approach would need to be passed to all value serializers, and the input and output bytes differed for Google wrappers (see test `.reference` file), so I decided to go for the current design.

jkuklis committed:
Pass the setting to buildFieldSerializer to fix undeclared

antonio2368 commented on src/Formats/ProtobufSerializer.cpp:
```diff
@@ -2280,7 +2321,17 @@ namespace
+                        {
+                            if (google_wrappers_special_treatment && nullableGoogleWrapper())
+                            {
+                                auto * nullable_ser = reinterpret_cast<ProtobufSerializerNullable*>(info.field_serializer.get());
```
```suggestion
                                auto * nullable_ser = dynamic_cast<ProtobufSerializerNullable*>(info.field_serializer.get());
```

antonio2368 commented on src/Formats/ProtobufSerializer.cpp:
```diff
@@ -2362,6 +2413,16 @@ namespace
                 missing_columns_filler->addDefaults(mutable_columns, row_num);
         }
 
+        bool nullGoogleWrapper(size_t row_num)
+        {
+            return isGoogleWrapperField(parent_field_descriptor) && mutable_columns[0].get()->isNullAt(row_num);
+        }
+
+        bool nullableGoogleWrapper()
+        {
+            return isGoogleWrapperField(parent_field_descriptor) && mutable_columns[0].get()->isNullable();
+        }
```
Names of these functions are bit unclear.

antonio2368 replied:
It is used only in place, and you can do something like
```
const bool is_google_wrapper = isGoogleWrapperField(parent_field_descriptor);
const bool is_column_nullable = mutable_columns[0].get()->isNullable();
if (is_google_wrapper && is_column_nullable)
  ...
```

jkuklis replied:
The trick is that for output I only have access to `mutable_columns` only if it is a Google wrapper (line 2238), so using `&&` in the helper function guards the `mutable_columns[0]` access in case this list is empty.
I can put it back as boolean conditions in the main parts, but these will need something like a nested if or an `&&` operation to prevent second part from firing for non-Google wrappers. I used `&&` earlier and I thought it looked rather ugly.

antonio2368 replied:
Okay, let's leave it then like it is.

antonio2368 reviewed (commented):
I'll go a bit more through this PR because there are some unclear things, but overall not bad.

antonio2368 commented:
@jkuklis glad you're back working on it, I was worried you dropped this 😄 

> The only difference in the current implementation is that Google wrappers carrying default values will be serialized as empty wrappers, as intended.

That sounds nice!

> Should the setting be true by default?

I would put it as false, i.e. opt-in for now. It isn't clear from anything that the google wrappers will interact with CH in such a way by default.  (+ there is another usecase described in the proto file "These types are useful
// for embedding primitives in the `google.protobuf.Any`") Let's play it safe for now.

For the other things I would need to check in more detail.

antonio2368 commented on src/Core/Settings.h:
```diff
@@ -639,6 +639,7 @@ class IColumn;
     M(MsgPackUUIDRepresentation, output_format_msgpack_uuid_representation, FormatSettings::MsgPackUUIDRepresentation::EXT, "The way how to output UUID in MsgPack format.", 0) \
     M(UInt64, input_format_max_rows_to_read_for_schema_inference, 100, "The maximum rows of data to read for automatic schema inference", 0) \
     M(Bool, input_format_json_read_bools_as_numbers, true, "Allow to parse bools as numbers in JSON input formats", 0) \
+    M(Bool, format_protobuf_google_wrappers_special_treatment, true, "Allow usage of Google value wrappers for regular values", 0) \
```
`input_format_protobuf_flatten_google_wrappers` and `output_format_protobuf_nullable_as_google_wrappers`.
What do you think?
Sorry for being so persistent with the naming, I would just like to be as specific as possible so the users are fully aware what's going on and so they have better control over parsing/serialising of the protobuf files.

jkuklis replied:
No worries, I appreciate the effort to make ClickHouse readable for the user ^^
To be completely explicit, even three settings could be introduced:
* `input_format_protobuf_flatten_google_wrappers`
* `input_format_protobuf_empty_google_wrappers_as_defaults`
* `output_format_protobuf_serialize_default_nullable_google_wrappers`
What do you think?

antonio2368 replied:
`output_format_protobuf_nullable_as_google_wrappers` I think is enough with a nice description, just to have setting name of reasonable length 😄 

> `input_format_protobuf_empty_google_wrappers_as_nullable_defaults` 

for this to work, the google wrapper needs to be flattened because `str.value` will be assigned to `str` column if there is a value. I.e. second setting needs to include the first one, and that's why it made sense to me to name it simply `input_format_protobuf_flatten_google_wrappers` (again, with a good description of the behaviour it produces).
The only reason to have it as a separate setting is if someone would like to flatten wrappers but not want the specific behaviour of null/default differentiation which I don't think will happen - please correct me if I'm wrong. 
Also, because it's an opt-in setting, we can easily change it someone actually will want it.

jkuklis replied:
I will introduce two settings then, one for input and one for output, both false by default.
So somebody might actually want to use only this flattening setting to start interpreting Google wrappers as regular columns, not necessarily nullable. This could be the case if someone had some component in their system which had to output Google wrappers, for example to ensure compatibility with some component other than ClickHouse, but wanted to interpret these as regular values in ClickHouse. The nullable aspect would not be important here.
But you are right the nullable part of the input depends on this flattening part. I also don't see why anyone would want to switch it off while keeping the flattening part, there is no legacy behaviour that could be affected.
For the output, the way this change could affect someone would be if they really expected a null when a default could be present, for example in tests on the message content or length, or when a 2-byte overhead per default nullable row was significant for the network (unlikely, I guess).

jkuklis committed:
Minor improvements

jkuklis committed:
Adding two settings in place of one, improvements to the test clarity

jkuklis committed:
Revert local change

jkuklis commented:
I added the requested changes, there are two settings now.

The test in the PR pipeline still fails, for the same weird reason as before:

> 2022-04-27 22:05:11 Error on processing query: Code: 36. DB::Exception: The format Protobuf requires a schema. The corresponding setting should be set: data for INSERT was parsed from query. (BAD_ARGUMENTS) (version 22.5.1.1)
2022-04-27 22:05:11 (query: INSERT INTO roundtrip_google_wrappers_02266 FORMAT Protobuf SETTINGS format_schema = '/usr/share/clickhouse-test/queries/0_stateless/format_schemas/02266_protobuf_format_google_wrappers:Message')

An earlier use of this schema for output works fine in the PR pipeline (though with `ProtobufSingle` format, not `Protobuf`), and locally the test passes. It would be quite tedious to debug this issue using only the PR pipeline, could you help me with that?

antonio2368 commented:
@mergifyio update

antonio2368 commented:
@jkuklis Can you please update with the latest master just to see if it happens again?

jkuklis committed:
Removing unused variable from test

jkuklis committed:
Update the test to have the correct order of FORMAT and SETTINGS

jkuklis committed:
Fix typo

jkuklis committed:
Reference output formatting corrections

jkuklis commented:
I rebased on master, and the test issue became clear: there was a change in the meantime, which requires the order for input to be `SETTINGS ... FORMAT ...` rather than the other way around like for output.

I fixed the test and it passes in the PR. Generally the tests look good, a couple of them failed, but I think these are irrelevant to the PR.
